language_service/src/hover.rs

```diff
@@ -25,67 +29,447 @@ pub(crate) fn get_hover(
+                let functors = if arrow.functors
+                    == hir::ty::FunctorSet::Value(hir::ty::FunctorSetValue::Empty)
+                {
+                    String::new()
+                } else {
+                    format!(" is {}", arrow.functors)
+                };
+                let arrow = match arrow.kind {
+                    hir::CallableKind::Function => "->",
+                    hir::CallableKind::Operation => "=>",
+                };
+                format!("({input} {arrow} {output}{functors})",)
+            }
+            hir::ty::Ty::Tuple(tys) => {
+                if tys.is_empty() {
+                    "Unit".to_owned()
+                } else {
+                    let elements = tys
+                        .iter()
+                        .map(|e| self.get_type_name_from_hir_ty(e))
+                        .collect::<Vec<_>>()
+                        .join(", ");
+                    format!("({elements})")
+                }
+            }
+            hir::ty::Ty::Udt(res) => match res {
+                hir::Res::Item(item_id) => {
+                    if let Some(item) = find_item(self.compilation, item_id) {
+                        match &item.kind {
+                            hir::ItemKind::Ty(ident, _) => ident.name.to_string(),
+                            _ => panic!("UDT has invalid resolution."),
                         }
+                    } else {
+                        "?".to_string()
+                    }
+                }
+                _ => panic!("UDT has invalid resolution."),
+            },
+            _ => ty.to_string(),
+        }
+    }
+}
+
+fn contents_from_hir_udt(name: &hir::Ident, _: &hir::ty::Udt) -> String {
+    let name = &name.name;
+
+    format!(
+        "```qsharp
+{name}
+```
+"
+    )
+}
+
+fn contents_from_ast_udt(name: &ast::Ident, def: &ast::TyDef) -> String {
+    let name = &name.name;
+    let def = ty_def_to_string(def);
+
+    format!(
+        "```qsharp
+{name}: {def}
+```
+"
+    )
+}
+
+fn ty_def_to_string(def: &ast::TyDef) -> String {
+    match &*def.kind {
+        ast::TyDefKind::Field(name, ty) => {
+            let ty = get_type_name_from_ast_ty(ty);
+            match name {
+                Some(name) => format!("{}: {ty}", name.name),
+                None => ty,
+            }
+        }
+        ast::TyDefKind::Paren(def) => ty_def_to_string(def),
+        ast::TyDefKind::Tuple(tys) => {
+            if tys.is_empty() {
+                "Unit".to_owned()
+            } else {
+                let elements = tys
+                    .iter()
+                    .map(|def| ty_def_to_string(def))
+                    .collect::<Vec<_>>()
+                    .join(", ");
+                format!("({elements})")
+            }
+        }
+    }
+}
+
+fn contents_from_name(name: &impl Display, ty_name: &String) -> String {
+    format!(
+        "```qsharp
```

Comment:
I don't love how the ```qsharp``` string is duplicated in so many functions. I think you can extract your inner formatting functions to separate methods (with `format_xyz()` names as I suggested earlier) and just have one place where you wrap them in markdown.

Reply:
I've refactored out a single function that does the markdown wrapping.